### PR TITLE
[WIP] Add rocFFT support

### LINKED ISSUE #4: Implement FFTs
https://github.com/ROCmSoftwarePlatform/rocFFT

### PR DESCRIPTION
**Pull request by antholzer**
Adds support for FFTs using [rocFFT](https://github.com/ROCmSoftwarePlatform/rocFFT).
I adapted the wrapper script from [CuArrays.jl/res/wrap/wrap.jl](https://github.com/JuliaGPU/CuArrays.jl/blob/master/res/wrap/wrap.jl) to create the basic C wrappings so it is also included.

Only partially working see: [issue #4 comment](https://github.com/jpsamaroo/ROCArrays.jl/issues/4#issuecomment-557165378)

Edit jps: closes #4

### CODE REVIEW

**Comment by antholzer** [comment]
Added hip wrapper for synchronization i.e. for the function defined [here](https://github.com/ROCm-Developer-Tools/HIP/blob/aaf31b6b96a71bf25d35c420e05c05271d22a403/include/hip/hcc_detail/hip_runtime_api.h#L783).
So it no longer needs my ugly workaround, but I am not sure if we want to use the HIP API.

**Comment by jpsamaroo** [comment]
This looks awesome, thanks! I'm :+1: to adding access to HIP routines via `ccall`, and using them in place of "native" HSA calls for now (or for the long term, if desired).

I'm happy with what I saw from a quick glance at the code, so just let me know once you'd like a full review and I'd be happy to do one :slightly_smiling_face:

**Comment by antholzer** [comment]
I think this should be more or less ready. @jpsamaroo you can review it and let me know what I should change.

The only cases that are still failing is the batched inverse and 3D complex (not inplace).
2D real got fixed in the newest release (2.10):
```
Test Summary:                  | Pass  Fail  Total
ROCArrays                      |  133    26    159
  ROCArrays External Libraries |  133    26    159
    Build Information          |    1            1
    Highlevel                  |    2            2
    Level 1 BLAS               |   12           12
    Level 2 BLAS               |    2            2
    FFT                        |  116    26    142
      T = Complex{Float64}     |   30     7     37
        1D                     |    3            3
        1D inplace             |    2            2
        2D                     |    3            3
        2D inplace             |    2            2
        Batch 1D               |    6            6
        3D                     |          3      3
        3D inplace             |    2            2
        Batch 2D (in 3D)       |    5     2      7
        Batch 2D (in 4D)       |    7     2      9
      T = Complex{Float32}     |   30     7     37
        1D                     |    3            3
        1D inplace             |    2            2
        2D                     |    3            3
        2D inplace             |    2            2
        Batch 1D               |    6            6
        3D                     |          3      3
        3D inplace             |    2            2
        Batch 2D (in 3D)       |    5     2      7
        Batch 2D (in 4D)       |    7     2      9
      T = Float32              |   28     6     34
        1D                     |    4            4
        2D                     |    4            4
        Batch 1D               |    4     2      6
        3D                     |    4            4
        Batch 2D (in 3D)       |    5     2      7
        Batch 2D (in 4D)       |    7     2      9
      T = Float64              |   28     6     34
        1D                     |    4            4
        2D                     |    4            4
        Batch 1D               |    4     2      6
        3D                     |    4            4
        Batch 2D (in 3D)       |    5     2      7
        Batch 2D (in 4D)       |    7     2      9
```

File: src/fft/fft.jl [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+const ROCFFT_INVERSE = false
+
+abstract type ROCFFTPlan{T,K,inplace} <: Plan{T} end
+
+Base.unsafe_convert(::Type{rocfft_plan}, p::ROCFFTPlan) = p.handle
+
+function unsafe_free!(plan::ROCFFTPlan)
+    rocfft_plan_destroy(plan.handle)
+    rocfft_execution_info_destroy(plan.execution_info)
+    # hsa_memory_free(plan.workarea.handle) |> check  # needed?
```
Comment (jpsamaroo) [inline src/fft/fft.jl]: Depends on if this is normally done in C++; if so, it might not hurt to have it (HSA will complain if we weren't supposed to free it anyway).

Comment (jpsamaroo) [reply src/fft/fft.jl]: Although, seems like it's a `ROCArray`, which frees its own memory via finalizer.

File: src/fft/fft.jl [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+                    ostrides .= real_strides
+                else
+                    odist = idist
+                end
+                if xtype == rocfft_transform_type_real_inverse
+                    idist = cdist
+                    istrides .= real_strides
+                end
+            else
+                # TODO this branch
```
Comment (jpsamaroo) [inline src/fft/fft.jl]: What about this branch is TODO?

File: src/fft/fft.jl [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+                        cem[ii] *= cdims[id]
+                        id += 1
+                    end
+                    @assert nem[ii] >= sz[ii]
+                end
+                if region[end] < length(xdims)
+                    ngaps += 1
+                end
+                # CUFFT represents batches by a single stride (_dist)
+                # TODO: ROCFFT can have multiple strides, so this not necessary?
```
Comment (jpsamaroo) [inline src/fft/fft.jl]: I suppose it's worth removing it and finding out :smile:

File: src/fft/fft.jl [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+            rocfft_plan_description_set_data_layout(description, in_array_type, out_array_type,
+                                                    C_NULL, C_NULL,
+                                                    length(istrides), istrides, idist,
+                                                    length(ostrides), ostrides, odist)
+            rocfft_plan_create(handle_ref, placement, xtype, precision, nrank, sz, batch, description)
+            rocfft_plan_description_destroy(description)
+        end
+    end
+    rocfft_plan_get_work_buffer_size(handle_ref[], worksize_ref)
+    # TODO allow empty array in ../array.jl
```
Comment (jpsamaroo) [inline src/fft/fft.jl:222]: I can come back to this once we have GPUArray support (which should fix this).

File: src/fft/fft.jl [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+    Xcopy = copy(X)
+    rocfft_execute(plan, [Xcopy.handle,], [Y.handle,], plan.execution_info)
+end
+
+
+function LinearAlgebra.mul!(y::ROCArray{Ty}, p::ROCFFTPlan{T,K,false}, x::ROCArray{T}) where {T,Ty,K}
+    assert_applicable(p, x, y)
+    unsafe_execute!(p, x, y)
+    # 0 is default stream
+    hipStreamSynchronize(Ptr{Cvoid}(UInt64(0)))
```
Comment (jpsamaroo) [inline src/fft/fft.jl:360]: I'm not super confident this will scale well, since there's no guarantee other operations (like those launched as kernels from AMDGPUnative) will use the same queue; we should remember to return to this at some point (not this PR).

File: wrap/patches/.dummy [hunk 6]
```diff
@@ -0,0 +1 @@
+put patches here
```
Comment (jpsamaroo) [inline wrap/patches/.dummy]: Is this file still needed?

**Review by jpsamaroo — APPROVED** [review]
Looks great! Thanks for all of the hard work :smile:

**Comment by jpsamaroo** [comment]
Any theories on why some of our tests are failing when they shouldn't be? Is it potentially an issue of numerical accuracy or something like that?

File: src/fft/fft.jl (hunk 1 above)
Comment (antholzer) [reply src/fft/fft.jl]: Forgot to remove it (HSA was complaining). I added it initially, since over at CuArrays they are freeing the workspace, but they have some custom memory management.

File: src/fft/fft.jl (hunk 2 above)
Comment (antholzer) [reply src/fft/fft.jl]: Deleted it, was a leftover.

File: src/fft/fft.jl (hunk 3 above)
Comment (antholzer) [reply src/fft/fft.jl]: It actually seems to check if the batch dimensions are sequential (FFTW supports non-sequential, but CuFFT not). I just tested it and rocFFT also does not support it, so we should leave it for now.

**Comment by antholzer** [comment]
> Any theories on why some of our tests are failing when they shouldn't be? Is it potentially an issue of numerical accuracy or something like that?

It is not numerical accuracy (errors are quite large). I recreated and example in C++ and posted an issue at [rocFFT#273](https://github.com/ROCmSoftwarePlatform/rocFFT/issues/273).
Maybe I'm just using the wrong parameters/options, but I doubt it (unless I misunderstood the documentation).

**Comment by jpsamaroo** [comment]
Thanks!
